### ISSUE TITLE
Thread safety of IAsyncContext<T>

### ISSUE THREAD

**Issue by KalleOlaviNiemitalo** [dotnet/extensions#4623]
In the Microsoft.Extensions.AsyncState library, the AsyncState.EnsureCount method expands the List\<object?> if that doesn't yet have enough elements.  This is called by the implementations of IAsyncState.TryGet, IAsyncState.Get, and IAsyncState.Set.  Those are in turn called by the implementation of IAsyncContext\<T>.

Expanding a List\<T\> is not a thread-safe operation.  If two threads are calling  IAsyncContext\<T>.Set in parallel with different type arguments for T, and one of those calls causes the list to be expanded, then the other update may be lost.

[IFeatureCollection](https://learn.microsoft.com/dotnet/api/microsoft.aspnetcore.http.features.ifeaturecollection?view=aspnetcore-7.0) documentation does not mention thread safety, and its default implementation [FeatureCollection](https://github.com/dotnet/aspnetcore/blob/ffa0a028464e13d46aaec0c5ad8de0725a4d5aa5/src/Extensions/Features/src/FeatureCollection.cs) uses a thread-unsafe Dictionary\<Type, Object>, so the Microsoft.AspNetCore.AsyncState.AsyncContextHttpContext\<T> implementation doesn't seem thread-safe for parallel Set, either.

Perhaps this kind of thread-unsafety is expected if the library is offered as an alternative to using HttpContext.Features directly.  It is unexpected for developers switching from AsyncLocal\<T>, though.  Especially because _reading_ the async context can also expand the list so parallel reads are not formally safe even if nothing is calling Set.

**Comment by davidfowl** [comment]
Here are my async local rules https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#asynclocalt

**Comment by mobratil** [comment]
@KalleOlaviNiemitalo I think you're right. `AsyncState` uses holder `AsyncStateHolder` with `List<object?>` collection. Write operations on that collection are not thread safe including `EnsureCapacity()` method. If the async flow is shared between multiple threads the `List<>` might become corrupted. I think we need to fix the holder object to hold thread safe collection.

**Comment by mobratil** [comment]
You’re right about the thread safety concerns. I believe using ConcurrentDictionary<> instead of List<> could be a solid solution to avoid these issues. However, there might be performance implications. I’m planning to create another PR to address this. I’m open to suggestions if anyone has alternative ideas or considerations.

**Comment by mobratil** [comment]
I created a PR for Microsoft.Extensions.AsyncState that uses ConcurrentDictionary<,> internally. This should make it finally thread safe and it is in line with async local rules (https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#asynclocalt) posted by @davidfowl 

The Microsoft.AspNetCore.AsyncState.AsyncContextHttpContext<T> implementation still uses FeatureCollection which is not thread safe. In my opinion this component can hardly be made thread safe. At least I'm not aware of any simple solution.
The state in AsyncContextHttpContext is bound to the lifetime of the request. If you span threads inside the request and share the data with those threads, the tricky situation can happen. When the request finishes and the threads are still running, how the behavior should look like? I think we should keep Microsoft.AspNetCore.AsyncState.AsyncContextHttpContext<T> as is unless I'm missing something.

**Comment by geeknoid** [comment]
Yeah, let's go with the documentation route...
